File: sdk/php/FounderOS.php
<?php

declare(strict_types=1);

namespace FounderOS;

/**
 * FounderOS analytics SDK — PHP (server-side).
 *
 * Buffers product events and flushes them in batches to the track-event edge
 * function. Authenticates with an `fos_` API key (Integrations → API Keys); the
 * workspace is resolved from the key, so only the project id is required.
 *
 * Because PHP is request-scoped, there is no background thread: events are
 * flushed when the buffer fills, when you call flush(), or automatically on
 * script shutdown (a shutdown handler is registered in the constructor).
 *
 * Usage:
 *   $fos = new \FounderOS\FounderOS([
 *       'host'       => 'https://xxxx.supabase.co',
 *       'project_id' => '<project-uuid>',
 *       'api_key'    => 'fos_...',
 *   ]);
 *   $fos->track('signup', 'user@example.com', ['plan' => 'pro']);
 *   $fos->flush();
 *
 * Requires the cURL extension (ships with most PHP installs).
 */
final class FounderOS
{
    private string $host;
    private string $projectId;
    private ?string $apiKey;
    private ?string $workspaceId;
    private ?string $anonKey;
    private int $batchSize;
    private bool $debug;
    private ?string $distinctId = null;

    /** @var array<int, array<string, mixed>> */
    private array $queue = [];

    /** @param array<string, mixed> $config */
    public function __construct(array $config)
    {
        if (empty($config['host']) || empty($config['project_id'])) {
            throw new \InvalidArgumentException('FounderOS: host and project_id are required');
        }
        $this->host        = rtrim((string) $config['host'], '/');
        $this->projectId   = (string) $config['project_id'];
        $this->apiKey      = isset($config['api_key']) ? (string) $config['api_key'] : null;
        $this->workspaceId = isset($config['workspace_id']) ? (string) $config['workspace_id'] : null;
        $this->anonKey     = isset($config['anon_key']) ? (string) $config['anon_key'] : null;
        $this->batchSize   = (int) ($config['batch_size'] ?? 20);
        $this->debug       = (bool) ($config['debug'] ?? false);

        if ($this->apiKey === null && ($this->workspaceId === null || $this->anonKey === null)) {
            throw new \InvalidArgumentException(
                'FounderOS: provide api_key (server) or workspace_id + anon_key'
            );
        }

        // Flush whatever is left when the request ends.
        register_shutdown_function([$this, 'flush']);
    }

    /** Associate subsequent events with a user. */
    public function identify(string $distinctId, ?array $properties = null): void
    {
        $this->distinctId = $distinctId;
        if ($properties !== null) {
            $this->track('$identify', $distinctId, $properties);
        }
    }

    /**
     * Queue an event. Sent on the next flush (or immediately when the buffer
     * reaches batch_size).
     *
     * @param array<string, mixed>|null $properties
     */
    public function track(
        string $eventName,
        ?string $distinctId = null,
        ?array $properties = null,
        ?string $occurredAt = null
    ): void {
        $this->queue[] = [
            'event_name'  => $eventName,
            'distinct_id' => $distinctId ?? $this->distinctId,
            'properties'  => $properties ?? new \stdClass(),
            'occurred_at' => $occurredAt ?? gmdate('Y-m-d\TH:i:s\Z'),
        ];
        if (count($this->queue) >= $this->batchSize) {
            $this->flush();
        }
    }

    /** Send all queued events now. */
    public function flush(): void
    {
        if (count($this->queue) === 0) {
            return;
        }
        $batch = $this->queue;
        $this->queue = [];
        try {
            $this->post('track-event', [
                'project_id'   => $this->projectId,
                'workspace_id' => $this->workspaceId,
                'batch'        => $batch,
            ]);
            $this->log('flushed ' . count($batch) . ' events');
        } catch (\Throwable $e) {
            // Re-queue so events aren't lost within this request.
            $this->queue = array_merge($batch, $this->queue);
            $this->log('flush failed, re-queued: ' . $e->getMessage());
        }
    }

    /** @param array<string, mixed> $body */
    private function post(string $fn, array $body): void
    {
        $headers = ['Content-Type: application/json'];
        if ($this->apiKey !== null) {
            $headers[] = 'Authorization: Bearer ' . $this->apiKey;
        }
        if ($this->anonKey !== null) {
            $headers[] = 'apikey: ' . $this->anonKey;
        }

        $ch = curl_init($this->host . '/functions/v1/' . $fn);
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER     => $headers,
            CURLOPT_POSTFIELDS     => json_encode($body, JSON_UNESCAPED_SLASHES),
            CURLOPT_TIMEOUT        => 10,
        ]);
        $resp   = curl_exec($ch);
        $status = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $err    = curl_error($ch);
        curl_close($ch);

        if ($resp === false) {
            throw new \RuntimeException("$fn request failed: $err");
        }
        if ($status < 200 || $status >= 300) {
            throw new \RuntimeException("$fn $status: " . substr((string) $resp, 0, 200));
        }
    }

    private function log(string $msg): void
    {
        if ($this->debug) {
            error_log('[founderos] ' . $msg);
        }
    }
}

File: sdk/python/founderos.py
"""FounderOS analytics SDK — Python (server-side).

Buffers product events and flushes them in batches to the track-event edge
function. Authenticates with an ``fos_`` API key (issued in Integrations → API
Keys); the workspace is resolved from the key, so only ``project_id`` is needed.

Usage::

    from founderos import FounderOS

    fos = FounderOS(
        host="https://xxxx.supabase.co",
        project_id="<project-uuid>",
        api_key="fos_...",
    )
    fos.track("signup", distinct_id="user@example.com", properties={"plan": "pro"})
    fos.track("feature_used", distinct_id="user@example.com",
              properties={"feature": "export"})
    fos.flush()        # or rely on the background flusher
    fos.shutdown()     # flush + stop the worker (call on exit)

Only depends on the standard library (urllib), so it drops into any environment.
"""

from __future__ import annotations

import atexit
import json
import threading
import time
import urllib.error
import urllib.request
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class FounderOS:
    def __init__(
        self,
        host: str,
        project_id: str,
        api_key: Optional[str] = None,
        workspace_id: Optional[str] = None,
        anon_key: Optional[str] = None,
        flush_interval: float = 5.0,
        batch_size: int = 20,
        max_queue: int = 10000,
        debug: bool = False,
    ) -> None:
        if not host or not project_id:
            raise ValueError("FounderOS: host and project_id are required")
        if not api_key and not (workspace_id and anon_key):
            raise ValueError(
                "FounderOS: provide api_key (server) or workspace_id + anon_key"
            )
        self.host = host.rstrip("/")
        self.project_id = project_id
        self.api_key = api_key
        self.workspace_id = workspace_id
        self.anon_key = anon_key
        self.batch_size = batch_size
        self.max_queue = max_queue
        self.debug = debug
        self._distinct_id: Optional[str] = None

        self._queue: List[Dict[str, Any]] = []
        self._lock = threading.Lock()
        self._stop = threading.Event()
        self._worker: Optional[threading.Thread] = None
        if flush_interval > 0:
            self._worker = threading.Thread(
                target=self._loop, args=(flush_interval,), daemon=True
            )
            self._worker.start()
        atexit.register(self.shutdown)

    # ── public API ──
    def identify(self, distinct_id: str, properties: Optional[Dict[str, Any]] = None) -> None:
        """Associate subsequent events with a user."""
        self._distinct_id = distinct_id
        if properties:
            self.track("$identify", distinct_id=distinct_id, properties=properties)

    def track(
        self,
        event_name: str,
        distinct_id: Optional[str] = None,
        properties: Optional[Dict[str, Any]] = None,
        occurred_at: Optional[str] = None,
    ) -> None:
        """Queue an event. Sent on the next flush (or immediately when full)."""
        event = {
            "event_name": event_name,
            "distinct_id": distinct_id or self._distinct_id,
            "properties": properties or {},
            "occurred_at": occurred_at or _now_iso(),
        }
        with self._lock:
            self._queue.append(event)
            over = len(self._queue) >= self.batch_size
            if len(self._queue) > self.max_queue:
                self._queue = self._queue[-self.max_queue:]
        if over:
            self.flush()

    def flush(self) -> None:
        """Send all queued events now."""
        with self._lock:
            if not self._queue:
                return
            batch = self._queue
            self._queue = []
        try:
            self._post("track-event", {
                "project_id": self.project_id,
                "workspace_id": self.workspace_id,
                "batch": batch,
            })
            self._log(f"flushed {len(batch)} events")
        except Exception as exc:  # re-queue so events aren't lost
            with self._lock:
                self._queue = (batch + self._queue)[: self.max_queue]
            self._log(f"flush failed, re-queued: {exc}")

    def shutdown(self) -> None:
        """Flush and stop the background worker. Idempotent."""
        if self._stop.is_set():
            return
        self._stop.set()
        if self._worker and self._worker.is_alive():
            self._worker.join(timeout=2.0)
        self.flush()

    # ── internals ──
    def _loop(self, interval: float) -> None:
        while not self._stop.wait(interval):
            self.flush()

    def _post(self, fn: str, body: Dict[str, Any]) -> None:
        data = json.dumps(body).encode("utf-8")
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        if self.anon_key:
            headers["apikey"] = self.anon_key
        req = urllib.request.Request(
            f"{self.host}/functions/v1/{fn}", data=data, headers=headers, method="POST"
        )
        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                resp.read()
        except urllib.error.HTTPError as exc:
            detail = exc.read().decode("utf-8", "ignore")[:200]
            raise RuntimeError(f"{fn} {exc.code}: {detail}") from None

    def _log(self, msg: str) -> None:
        if self.debug:
            print(f"[founderos] {msg}")
